### ISSUE TITLE
Plans for Package

### ISSUE THREAD

**Issue by nstarman** [data-apis/array-api-typing#1]
Here's a list of my suggestions for this package. @jorenham LMK what you agree with.

Configuration/CI/CD:
- `pyproject.toml`, no `setup.py` or `setup.cfg`
- use `ruff` (configured in pyproject.toml)
- use `uv`, including in [CI](https://docs.astral.sh/uv/guides/integration/github/) 
- build-system is `hatch` and `hatch-vcs`
- use `pre-commit`
- check with mypy and pyright, the most common type checkers. 

Docs:
- same look and feel as the main docs.
- markdown

Package layout:
- src-layout
- use walled-garden separation between public and private API for easy adherence to PEP8.

```
src/
    array_api_typing/
        __init__.py  # defines the public API
        _src/  # private API walled garden. Everything here is private and easily worked on by devs w/out public API consequences.
            base.py  # named whatever we want. no pre-fix underscores.
            core.py  # named whatever we want. no pre-fix underscores.
```

Package Contents:
- `Array`. Let's start (pre-release) by not making it parametric, then see what we can do well, e.g. lessons from optype.
    - I would actually like to have 2 classes a "minimal" and "full", where the minimal is only

```python
@runtime_checkable
class MinimalArray(Protocol):
    def __array_namespace__(self, ...) -> Namespace: ...
```

and the full version contains all the attributes and methods.

- `Namespace`
- `DType`, `Device`, etc.

**Comment by jorenham** [comment]
Oh haha I just pushed some initial stuff right now

**Comment by jorenham** [comment]
I certainly agree with `uv` + `hatch`, `pre-commit`, and markdown for docs. 

And to that I'd like to add `ruff`, [`sp-repo-review`](https://pypi.org/project/sp-repo-review/), `tox` (maybe with [`tox-uv`](https://github.com/tox-dev/tox-uv)), and some markdown linter like [markdownlint](https://github.com/DavidAnson/markdownlint), as well.

**Comment by nstarman** [comment]
Ruff is on the list 😄. Agreed on sp-repo-review, I use that frequently.
Can we use `nox` instead of `tox`?
👍 to a markdown linter.
Also `pytest` as the test framework
And [`sybil`](https://sybil.readthedocs.io/en/latest/) to enable documentation tests as part of the CI

**Comment by nstarman** [comment]
> > * check with mypy and pyright, the most common type checkers.
> 
> I prefer `basedmypy` instead `mypy`, and `basedpyright` instead of `pyright`. Both come with additional (more strict) checks, better defaults, and fewer bugs (especially in the case of mypy, see https://github.com/erictraut/mypy_issues).
> 
> I've been using both for a while now in all of my (typed) projects, and never had any problems with backwards-compatibility or something.

Great! I'm totally down to use stricter checkers. But we should have the most used ones as part of the CI. We can't force users, whom given the scope of the array-api we hope will be a large and varied audience, to use specific type checkers.

**Comment by nstarman** [comment]
@jorenham What do you think of the rest of the suggestions? Just compiling the full agreed-up list up top.

**Comment by jorenham** [comment]
> Can we use `nox` instead of `tox`?

I'm not very familiar with `nox`, but last time I checked it looked rather limited, and is was missing `--parallel`, which was the deal breaker for me at that time. But perhaps things changed since then?

> Also pytest as the test framework

Yes please. 


> And [`sybil`](https://sybil.readthedocs.io/en/latest/) to enable documentation tests as part of the CI

Hmm first time I heard of it. I've used [pytest-doctestplus](https://github.com/scientific-python/pytest-doctestplus) in [Lmo](https://github.com/jorenham/lmo/), which sounds similar. Any idea how these two differ?


> We can't force users, whom given the scope of the array-api we hope will be a large and varied audience, to use specific type checkers.

100% agree

I think it's also important to explicitly state somewhere that we follow the [official typing spec](https://typing.readthedocs.io/en/latest/spec/special-types.html), knowing that mypy doesn't always follow it (and has >1000 confirmed bugs at the moment).

**Comment by jorenham** [comment]
> @jorenham What do you think of the rest of the suggestions? Just compiling the full consensus list up top.

Maybe we should discuss the runtime contents of the package in another issue, and keep this one focussed on the tooling and CI stuff?

**Comment by nstarman** [comment]
> > Can we use `nox` instead of `tox`?
> 
> I'm not very familiar with `nox`, but last time I checked it looked rather limited, and is was missing `--parallel`, which was the deal breaker for me at that time. But perhaps things changed since then?

I don't think it has `—parallel` yet, but is that a problem for this light-weight package?
Nox also makes coordinating a suite of different CI options, e.g. `test-mypy`, `test-basedmpypy` super easy.

> > Also pytest as the test framework
> 
> Yes please.

🎉 

> > And [`sybil`](https://sybil.readthedocs.io/en/latest/) to enable documentation tests as part of the CI
> 
> Hmm first time I heard of it. I've used [pytest-doctestplus](https://github.com/scientific-python/pytest-doctestplus) in [Lmo](https://github.com/jorenham/lmo/), which sounds similar. Any idea how these two differ?

pytest-doctestplus used to be in Astropy (which I'm still in). I've used both extensively and can say that `sybil` is more intuitive, customizable, and powerful.

> I think it's also important to explicitly state somewhere that we follow the [official typing spec](https://typing.readthedocs.io/en/latest/spec/special-types.html), knowing that mypy doesn't always follow it (and has >1000 confirmed bugs at the moment).

Totally agree.

**Comment by nstarman** [comment]
> When running on multiple python versions then it already helps. In our case, we're probably gonna want to include big-boy packages like tensorflow and cupy in the testing pipeline, which could slow things down a lot.

Sure, then let's start with tox. I'll experiment with `nox` at a later date.

**Comment by jorenham** [comment]
Oh and perhaps it won't really make a lot of sense to use it in our case, but in case that it does, then [`hypthesis`](https://hypothesis.readthedocs.io/en/latest/) could be a pretty neat testing framework (it has seamless pytest integration)

**Comment by jorenham** [comment]
> > When running on multiple python versions then it already helps. In our case, we're probably gonna want to include big-boy packages like tensorflow and cupy in the testing pipeline, which could slow things down a lot.
> 
> Sure, then let's start with tox. I'll experiment with `nox` at a later date.

Tox + [`tox-uv`](https://github.com/tox-dev/tox-uv) could also be used as a full fledged task runner, as noted here: https://github.com/jorenham/scipy-stubs/issues/216#issuecomment-2507339719

**Comment by jorenham** [comment]
and semver for versioning?

**Comment by nstarman** [comment]
💯

**Comment by jorenham** [comment]
and should we enfore a commit style, like [gitmoji](https://gitmoji.dev/) or something?

**Comment by nstarman** [comment]
> and should we enfore a commit style, like [gitmoji](https://gitmoji.dev/) or something?

Yes.

**Comment by jorenham** [comment]
> Although... given the date-versioning of the array-api...

good point 🤔

We probably don't want to limit outselves to a single array-api version, so I think we can just use our own independent versioning scheme?

**Comment by jorenham** [comment]
> Do you know if there's a way to integrate that with `commitizen`?

First time I heard of `commitizen` 🤷🏻 
I use the [vscode plugin](https://marketplace.visualstudio.com/items?itemName=seatonjiang.gitmoji-vscode) myself

**Comment by nstarman** [comment]
> We probably don't want to limit outselves to a single array-api version, so I think we can just use our own independent versioning scheme?

Spitballing. We probably want to commit to supporting, e.g. 2-versions back + most recent, of the array-api.
If we do the git version branch dev strategy (so we have `main`, `2023-12-11`, `2024-05-12` or whatever is the array-api)
Then we can open feature PRs against main and the other versions independently. We can also easily backport PRs using https://github.com/meeseeksmachine (used by matplotlib, ipython, etc) from `main` to the other branches.

**Comment by nstarman** [comment]
I use `commitizen` to just enforce the commit message style. It's very much like the VSCode plugin you mention, just applicable to other IDEs and devs (anyone trying to open a PR). You ok using that as to enforce gitmoji?

**Comment by nstarman** [comment]
https://github.com/adamchainz/blacken-docs for pre-commit?

**Comment by nstarman** [comment]
https://github.com/codespell-project/codespell or https://cspell.org for pre-commit?

**Comment by jorenham** [comment]
I think that the availability of [`__array_namespace_info__`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.info.__array_namespace_info__.html) is a very important one for us, but that's only in the spec since 2023... Because as far as I'm aware, that the only way for us to figure out what the type of a device is. 

So yee, maybe we do need to maintain different versions. And if we do, then it indeed makes sense to start with the earliest one that can be reasonably typed, and then work our way up. I'm not familiar with https://github.com/meeseeksmachine, but it sounds like something that could be interesting for `scipy-stubs` as well.

**Comment by jorenham** [comment]
> https://github.com/codespell-project/codespell or https://cspell.org for pre-commit?

I use that in all of my projects, but forgot to mention it, so yes please!

**Comment by nstarman** [comment]
> > https://github.com/codespell-project/codespell or https://cspell.org for pre-commit?
> 
> I use that in all of my projects, but forgot to mention it, so yes please!

Which? no personal preference.

**Comment by jorenham** [comment]
> > > https://github.com/codespell-project/codespell or https://cspell.org for pre-commit?
> > 
> > 
> > I use that in all of my projects, but forgot to mention it, so yes please!
> 
> Which? no personal preference.

What's the difference?

**Comment by jorenham** [comment]
> > > https://github.com/codespell-project/codespell or https://cspell.org for pre-commit?
> > 
> > 
> > I use that in all of my projects, but forgot to mention it, so yes please!
> 
> Which? no personal preference.

Me neitehr, I wouldn't mind a better alternative to codespell, as its kinda annoying to configure, and configuration can only be done globally.

**Comment by nstarman** [comment]
> I think that the availability of [`__array_namespace_info__`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.info.__array_namespace_info__.html) is a very important one for us, but that's only in the spec since 2023... Because as far as I'm aware, that the only way for us to figure out what the type of a device is.
> 
> So yea, maybe we do need to maintain different versions. And if we do, then it indeed makes sense to start with the earliest one that can be reasonably typed, and then work our way up. I'm not familiar with https://github.com/meeseeksmachine, but it sounds like something that could be interesting for `scipy-stubs` as well.

Re meeseeks, see https://github.com/astropy/astropy/pull/17220 for an example usage.

So then yes to the `main` + branch-per-prevsous-version format? Coordinate with meeseeks when backporting from main to previous versions. If we're starting with an older version and working up, then we'll commit to main and always backport. Then we release the older version from that branch and start on the newer version, committing to main and backporting to the next branch. We'll work our way up to the present!

**Comment by nstarman** [comment]
> > > > https://github.com/codespell-project/codespell or https://cspell.org for pre-commit?
> > > 
> > > 
> > > I use that in all of my projects, but forgot to mention it, so yes please!
> > 
> > 
> > Which? no personal preference.
> 
> Me neither, I wouldn't mind a better alternative to codespell, as its kinda annoying to configure, and configuration can only be done globally.

Totally agree. I use codespell because I've always used codespell. And I've always been irked by its poor configurability.

**Comment by jorenham** [comment]
> So then yes to the `main` + branch-per-prevsous-version format?

Yea sure, but I suppose that the details are a problem of future us at this point.

**Comment by jorenham** [comment]
> > > > > https://github.com/codespell-project/codespell or https://cspell.org for pre-commit?
> > > > 
> > > > 
> > > > I use that in all of my projects, but forgot to mention it, so yes please!
> > > 
> > > 
> > > Which? no personal preference.
> > 
> > 
> > Me neither, I wouldn't mind a better alternative to codespell, as its kinda annoying to configure, and configuration can only be done globally.
> 
> Totally agree. I use codespell because I've always used codespell. And I've always been irked by its poor configurability.

cspell looks very js-oriented. Does it support python?

**Comment by nstarman** [comment]
> cspell looks very js-oriented. Does it support python?

Yes, but I'm not sure how it stacks up against codespell, I think it's newer to the scene so has had less adoption by older major projects, so I don't have much reference besides the docs...

**Comment by jorenham** [comment]
let's stick to waht we know then. We don't want to risk becoming `array-api-typo` instead

**Comment by jorenham** [comment]
However, I just found this one, and it supports Python:
https://github.com/crate-ci/typos

**Comment by nstarman** [comment]
SGTM. https://github.com/crate-ci/typos/issues/361#issuecomment-1711720189 so we can use pyproject.toml

**Comment by nstarman** [comment]
https://github.com/adamchainz/blacken-docs for pre-commit?

**Comment by jorenham** [comment]
> https://github.com/adamchainz/blacken-docs for pre-commit?

is that the one that `repo-review` requires?

**Comment by jorenham** [comment]
Ah yes it is; I'm actually using it in `optype`:
https://github.com/jorenham/optype/blob/3b36fcf50a6fe1dbc730f5bd21f24e700bba9764/.pre-commit-config.yaml#L42-L47

So that's a 👍🏻

**Comment by nstarman** [comment]
Ah. Yes, I think so. Covered 👍

**Comment by nstarman** [comment]
Have we covered all the basics?

**Comment by jorenham** [comment]
I briefly tried mdformat + [mdformat-ruff](https://pypi.org/project/mdformat-ruff/) a while back, but it didn't live up to its expectations unfortunately

**Comment by jorenham** [comment]
> Have we covered all the basics?

Yea I believe so. We can go ahead and do the actual setup now 👍🏻

**Comment by jorenham** [comment]
The current `.editorconfig` is already black compatible FYI

**Comment by jorenham** [comment]
Do you mind if I force push to main to update the commit message in gitmoji style?

**Comment by jorenham** [comment]
(I could also undo the commit alltogether, and do it over with a PR)

**Comment by nstarman** [comment]
> (I could also undo the commit alltogether, and do it over with a PR)

Sure! A beautiful git history from the start ✨

**Comment by jorenham** [comment]
apparently at least one commit is required, so I made it an empty one instead

**Comment by nstarman** [comment]
@jorenham, what's the min version of the API we want to support?

**Comment by jorenham** [comment]
I guess 2021?

**Comment by nstarman** [comment]
What are the set of labels that we want? Maybe start with the full gitmoji set?

**Comment by jorenham** [comment]
> What are the set of labels that we want? Maybe start with the full gitmoji set? Good for searching Issues / PRs by label

I was just thinking about the labels. But yea, I guess we could try the gitmoji approach? That way we could even automate it if we wanted to

**Comment by nstarman** [comment]
SGTM!

**Comment by nstarman** [comment]
> I was just thinking about the labels. But yea, I guess we could try the gitmoji approach?

I added gitmoji labels.

**Comment by lucascolley** [comment]
> what's the min version of the API we want to support?

I would suggest working on the most recent version first, though

**Comment by nstarman** [comment]
In terms of backporting code across versions, it's much harder to work backwards than forwards.

**Comment by jorenham** [comment]
> > what's the min version of the API we want to support?
> 
> I would suggest working on the most recent version first, though

The stuff that will work for v2021 will probably also work with v2024, nut not the other way around.

**Comment by jorenham** [comment]
> > > what's the min version of the API we want to support?
> > 
> > 
> > I would suggest working on the most recent version first, though
> 
> The stuff that will work for v2021 will probably also work with v2024, nut not the other way around.

... and I guess what I'm trying to say with this; is that we'll *also* be atarting with the latest version if we start with the first version. 
Maybe I'm not making much sense when I say this, but the image that comes to mind is a something with layers, like a cake or something 🤷🏻.

But there's of course the chance that when we "arrive" at the latest version, we realize that we need to adjust something in the first "layer" in order for us to add the sprinkles on the top. But because we don't have to worry about backcomp (i.e. because of https://github.com/data-apis/array-api-typing/issues/1#issuecomment-2509146133), that's not a problem, so we could take bites out of the bottom of cake without issue (ok I'm hungry now).

**Comment by lucascolley** [comment]
I think I get you 😅 I just meant that the most important "end-goal" is compatibility with the latest version of the standard - feel free to do whatever makes sense to get there

**Comment by jorenham** [comment]
@lucascolley assuming that the array api versions are backwards compatible, then we'll always be compatible with the latest version this way 👌🏻

**Comment by nstarman** [comment]
It would also be great to get tox set up. Reproducing the CI is a bit challenging as a one-liner with uv.

``uv run --group=mypy --with "numpy==1.25.0" —active mypy --tb --no-incremental tests/integration/test_numpy1.pyi —pretty``
isn't quite the same as what's in the CI.

**Comment by jorenham** [comment]
I like `tox`: https://github.com/jorenham/optype/blob/4d180a49872afe3e84f30f29f2215e4350a6c5bf/pyproject.toml#L219-L311


Also, `uv run --group=mypy --with "numpy==1.25.0"` doesn't work, because mypy ignores layered dependencies
